presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDictionaryStreamReader.java

```diff
@@ -225,23 +266,67 @@ private void openRowGroup()
         rowGroupOpen = true;
     }
 
-    private static void readDictionary(@Nullable ByteArrayStream dictionaryDataStream, int dictionarySize, int[] dictionaryLength, Slice[] dictionary)
+    private static void readDictionary(@Nullable ByteArrayStream dictionaryDataStream, int dictionarySize, int[] dictionaryLength, Slice[] dictionary, boolean useSharedByteArray)
             throws IOException
     {
-        // build dictionary slices
-        for (int i = 0; i < dictionarySize; i++) {
-            int length = dictionaryLength[i];
-            if (length == 0) {
-                dictionary[i] = Slices.EMPTY_SLICE;
+        if (useSharedByteArray) {
+            int startIndex = 0;
+            while (startIndex < dictionarySize) {
+                // sum lengths up to CHUNK_BYTES
+                int totalLength = 0;
+                int endIndex = startIndex;
+                while (endIndex < dictionarySize) {
+                    int newTotalLength = totalLength + dictionaryLength[endIndex];
+                    if (newTotalLength > CHUNK_BYTES) {
+                        break;
```

Comment:
this could infinite loop if a single dictionary entry is larger than `CHUNK_BYTES`. This check should probably be `totalLength > 0 && newTotalLength > CHUNK_BYTES`